lib/RestructureCFG/EnforceSingleExitPass.cpp

```diff
@@ -0,0 +1,171 @@
+/// header, where a similar operation is done in order to compute the
+/// `PostDominatorTree` for a generic CFG, which can contain multiple exit
+/// blocks, and infinite loops too.
+class EnforceSingleExitPassImpl {
+  Function &F;
+
+public:
+  EnforceSingleExitPassImpl(Function &F) : F(F) {}
+
+public:
+  bool run() {
+    llvm::SmallVector<BasicBlock *> Roots;
+
+    // 1: Find all the trivial roots of the graph, i.e., the blocks which do not
+    //    have any successors on the `ScopeGraph`
+    for (BasicBlock &BB : F) {
+      if (not hasScopeGraphSuccessors(&BB)) {
+        Roots.push_back(&BB);
+      }
+    }
+
+    // 2: Find all the non trivial roots, i.e., those which are in infinite
+    //    loops. To do this, we iterate over all the SCCs present in the
+    //    `ScopeGraph` (mind the `GraphT` type passed to the `scc_iterator`),
+    //    and search for components that do not have any exiting edges.
+    using GraphT = Scope<llvm::BasicBlock *>;
+    BasicBlock *Entry = &F.getEntryBlock();
+    for (scc_iterator<GraphT> I = scc_begin(Scope(Entry)),
+                              IE = scc_end(Scope(Entry));
+         I != IE;
+         ++I) {
+
+      // We skip any SCC which doesn't compose a cycle
+      if (not I.hasCycle()) {
+        continue;
+      }
+
+      // We transpile the nodes composing the SCC in a set to have faster lookup
+      llvm::SmallSet<BasicBlock *, 4> SCCNodes;
+      for (auto *BB : *I) {
+        SCCNodes.insert(BB);
+      }
+
+      // We search for any block in the SCC that have an exiting edge (on the
+      // `ScopeGraph` too, mind the `GT` parameter)
+      bool HasExit = false;
+      for (auto *BB : *I) {
+        using GT = llvm::GraphTraits<Scope<llvm::BasicBlock *>>;
+        auto Successors = make_range(GT::child_begin(BB), GT::child_end(BB));
+        for (auto Successor : Successors) {
+          if (!SCCNodes.contains(Successor)) {
+            HasExit = true;
+            break;
+          }
+        }
+
+        // We are not interested in SCCs that already have an exit, also on the
+        // `ScopeGraph`
+        if (HasExit)
+          break;
+      }
+
+      // If we found an SCC having an exiting edge on the `ScopeGraph`, we move
+      // on to the next SCC
+      if (HasExit)
+        continue;
+
+      // We elect the "furthest away" node, along some DFS, for the elected SCC.
+      // Due to the internal behavior of the `scc_iterator`, we can elect the
+      // first node in each `SCC` as the furthest away node on the DFS used
+      // during the `scc_iterator` itself. This would
+      BasicBlock *LastSCCNode = *I->begin();
+      Roots.push_back(LastSCCNode);
+    }
+
+    // We should have collected at least one root
+    revng_assert(Roots.size() != 0);
+
+    // 3: We add an additional `sink_block` where to redirect all the roots
+    //    found only if we have more than one root
+    if (Roots.size() <= 1) {
+      return false;
+    }
```

Comment:
The new block should be added in all the cases where all detected exits are non-trivial.

In all the other cases, there's at least one eligible trivial exits that gets marked as "one true exit".

Reply:
Ok, I adopted the new strategy that you described, there are a couple of changes around due to this, like keeping two `vector`s for the _trivial_ and _non trivial_ exits https://github.com/revng/revng/blob/98f7b2d8804ed56f760546885bbbff6c72933843/lib/RestructureCFG/EnforceSingleExitPass.cpp#L69, let me know if you approve them

Reply:
Yeah it makes sense.